In [320]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

import ast

# 1 Charging Set

In [398]:
charging_set = pd.read_csv("data/charging_sessions.csv")

## 1.1 Overview of Raw Data

In [399]:
print("columns:")
print(charging_set.columns)
print("-------------------------------------------------------------------------")
print("info():")
print(charging_set.info())
print("-------------------------------------------------------------------------")
print("describe():")
print(charging_set.describe())

columns:
Index(['Unnamed: 0', 'id', 'connectionTime', 'disconnectTime',
       'doneChargingTime', 'kWhDelivered', 'sessionID', 'siteID', 'spaceID',
       'stationID', 'timezone', 'userID', 'userInputs'],
      dtype='object')
-------------------------------------------------------------------------
info():
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone    

In [400]:
charging_set.head(10)

Unnamed: 0                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
5           5  5e23b149f9af8b5fe4b973d4  2020-01-02 14:00:13+00:00   
6           6  5e23b149f9af8b5fe4b973d5  2020-01-02 14:09:14+00:00   
7           7  5e23b149f9af8b5fe4b973d6  2020-01-02 14:17:32+00:00   
8           8  5e23b149f9af8b5fe4b973d7  2020-01-02 14:25:38+00:00   
9           9  5e23b149f9af8b5fe4b973d8  2020-01-02 14:27:40+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
5  2020-01-02 23:00:41+00:00  2020-01-02 20:13:21+00:00        13.370   
6  2020-01-03 01:57:58+00:00  2020-01-03 00:59:21+00:00        43.477   
7  2020-01-03 00:44:54+00:00  2020-01-02 20:22:40+00:00        32.441   
8  2020-01-02 19:36:05+00:00  2020-01-02 18:38:39+00:00        13.264   
9  2020-01-02 20:01:48+00:00  2020-01-02 19:17:46+00:00        13.285   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
5  1_1_194_821_2020-01-02 14:00:05.241573       1  AG-1F13  1-1-194-821   
6  1_1_178_817_2020-01-02 14:09:14.451827       1  AG-1F09  1-1-178-817   
7  1_1_191_804_2020-01-02 14:17:32.329535       1  AG-4F34  1-1-191-804   
8  1_1_194_826_2020-01-02 14:25:37.578692       1  AG-1F11  1-1-194-826   
9  1_1_178_823_2020-01-02 14:27:39.582337       1  AG-1F08  1-1-178-823   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   
5  America/Los_Angeles  3519.0   
6  America/Los_Angeles   933.0   
7  America/Los_Angeles  3737.0   
8  America/Los_Angeles   419.0   
9  America/Los_Angeles   651.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
5  [{'WhPerMile': 600, 'kWhRequested': 24.0, 'mil...  
6  [{'WhPerMile': 385, 'kWhRequested': 65.45, 'mi...  
7  [{'WhPerMile': 300, 'kWhRequested': 45.0, 'mil...  
8  [{'WhPerMile': 400, 'kWhRequested': 40.0, 'mil...  
9  [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...

## 1.2 Remove Unnecessary Columns

We remove the columns that have no additional value to our dataset, either because the column makes no sense (e.g. "Unnamed 0") or  because the data only holds one information that was already given by us (e.g. "timezone" only has "America/Los_Angeles").

In [401]:
charging_set.drop('Unnamed: 0', axis=1, inplace=True)
charging_set.drop('timezone', axis=1, inplace=True)

## 1.3 Splitting Multi-Info Columns

With the first look on the head of the dataset the column "userInputs" stands out since the data in that column is a dictionary. We split this column into separate columns first.

In [402]:
def parse_user_inputs(row):
    """
    This function parses one row of data that is a python object, like a dictionary in this case 
    
    param row : string representation of python object like a dictionary, or NaN.
    
    :return: first dictionary from the parsed list if valid, or an empty dictionary otherwise.
    """
    try:
        # Parse the row as a Python object (e.g., list of dicts)
        parsed = ast.literal_eval(row) if pd.notnull(row) else []
        return parsed[0] if isinstance(parsed, list) and len(parsed) > 0 else {}
    except (ValueError, SyntaxError):
        return {}

# Parse 'userInputs' column in charging_set
charging_set['userInputs'] = charging_set['userInputs'].apply(parse_user_inputs)

# Specify the desired features to extract
desired_features = [
    'WhPerMile', 'kWhRequested', 'milesRequested',
    'minutesAvailable', 'modifiedAt', 'paymentRequired',
    'requestedDeparture'
]

# Extract desired features and create new columns
for feature in desired_features:
    charging_set[feature] = charging_set['userInputs'].apply(
        lambda x: x.get(feature) if isinstance(x, dict) else None
    )
charging_set = charging_set.drop(columns=['userInputs'])
charging_set

id             connectionTime  \
0      5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1      5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2      5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3      5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4      5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...                         ...                        ...   
66445  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   
66449  5d574ad2f9af8b4c10c03656  2019-07-31 19:21:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   
66449  2019-07-31 22:00:04+00:00  2019-07-31 20:51:34+00:00         0.897   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697       1  AG-1F09  1-1-178-817   

       userID  WhPerMile  kWhRequested  milesRequested  minutesAvailable  \
0       194.0      250.0         25.00           100.0             463.0   
1      4275.0      280.0         70.00           250.0             595.0   
2       344.0      400.0          8.00            20.0              60.0   
3      1117.0      400.0          8.00            20.0              65.0   
4       334.0      400.0         16.00            40.0             504.0   
...       ...        ...           ...             ...               ...   
66445   393.0      240.0         31.20           130.0             355.0   
66446   220.0      333.0          6.66            20.0             455.0   
66447  1974.0      333.0         19.98            60.0             184.0   
66448   942.0      275.0         22.00            80.0              60.0   
66449     NaN        NaN           NaN             NaN               NaN   

                          modifiedAt paymentRequired  \
0      Thu, 02 Jan 2020 13:09:39 GMT            True   
1      Thu, 02 Jan 2020 13:37:11 GMT            True   
2      Thu, 02 Jan 2020 13:57:17 GMT            True   
3      Thu, 02 Jan 2020 14:00:03 GMT            True   
4      Thu, 02 Jan 2020 14:00:13 GMT            True   
...                              ...             ...   
66445  Wed, 31 Jul 2019 18:08:23 GMT            True   
66446  Wed, 31 Jul 2019 18:41:02 GMT            T

## 1.3 Duplicates

In [403]:
# check number of duplicates
n_duplicates = charging_set.duplicated().sum()
print("sum of duplicates:", n_duplicates)

# remove all duplicates
charging_set.drop_duplicates(inplace=True)
(charging_set.head(5))

sum of duplicates: 1413


id             connectionTime  \
0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

   userID  WhPerMile  kWhRequested  milesRequested  minutesAvailable  \
0   194.0      250.0          25.0           100.0             463.0   
1  4275.0      280.0          70.0           250.0             595.0   
2   344.0      400.0           8.0            20.0              60.0   
3  1117.0      400.0           8.0            20.0              65.0   
4   334.0      400.0          16.0            40.0             504.0   

                      modifiedAt paymentRequired  \
0  Thu, 02 Jan 2020 13:09:39 GMT            True   
1  Thu, 02 Jan 2020 13:37:11 GMT            True   
2  Thu, 02 Jan 2020 13:57:17 GMT            True   
3  Thu, 02 Jan 2020 14:00:03 GMT            True   
4  Thu, 02 Jan 2020 14:00:13 GMT            True   

              requestedDeparture  
0  Thu, 02 Jan 2020 20:51:54 GMT  
1  Thu, 02 Jan 2020 23:31:50 GMT  
2  Thu, 02 Jan 2020 14:56:35 GMT  
3  Thu, 02 Jan 2020 15:04:58 GMT  
4  Thu, 02 Jan 2020 22:24:01 GMT

## 1.4 Missing Data

In [404]:
# check number of none-values
nan_counts = charging_set.isna().sum()
nan_counts_sorted = nan_counts.sort_values(ascending=False)

for column, count in nan_counts_sorted.items():
    print(f"Feature: {column}, Count of NaN-Values: {count}")

Feature: requestedDeparture, Count of NaN-Values: 17215
Feature: milesRequested, Count of NaN-Values: 17215
Feature: paymentRequired, Count of NaN-Values: 17215
Feature: modifiedAt, Count of NaN-Values: 17215
Feature: minutesAvailable, Count of NaN-Values: 17215
Feature: WhPerMile, Count of NaN-Values: 17215
Feature: kWhRequested, Count of NaN-Values: 17215
Feature: userID, Count of NaN-Values: 17215
Feature: doneChargingTime, Count of NaN-Values: 4087
Feature: id, Count of NaN-Values: 0
Feature: stationID, Count of NaN-Values: 0
Feature: siteID, Count of NaN-Values: 0
Feature: spaceID, Count of NaN-Values: 0
Feature: kWhDelivered, Count of NaN-Values: 0
Feature: disconnectTime, Count of NaN-Values: 0
Feature: connectionTime, Count of NaN-Values: 0
Feature: sessionID, Count of NaN-Values: 0


We observe a large number of NaN-values in the columns that were separated from the column \"userInputs\". It is very noticeable that the number of NaN-values of the \"userInputs\"-columns is the same as the number of the NaN-values in the \"userID\"-column. Accordingly, we assume that the NaN-values in the \"userInputs\"-columns originate from users not being registered. In the following, we check if our assumption is correct:

In [405]:
# list of userInputs columns
user_inputs = ["requestedDeparture", "milesRequested", "paymentRequired", "modifiedAt", "minutesAvailable", "WhPerMile", "kWhRequested"]

def check_user_inputs_nan_condition(user_inputs_columns):
    """
    This function checks if userID is NaN for all NaN records in the userInputs columns.
    
    :param user_inputs_columns: list of user input columns that were stored in the former "userInputs"-column
    :return: True if no violations are found, False if a userID exists with NaN in the corresponding userInput column.
    """
    # iterate over each row and compare userID-value to userInputs-values
    for idx, row in charging_set.iterrows():
        user_id_value = row["userID"]

        for user_input_column in user_inputs_columns:
            user_input_value = row[user_input_column]

            # if userID is not NaN but the userInputs-column is NaN, return false
            if (pd.isna(user_input_value) and not pd.isna(user_id_value)) or (not pd.isna(user_input_value) and pd.isna(user_id_value)):
                return False

    # if no violation of condition is found, return true
    return True

# check if unregistered users are the reason for NaN-values in userInputs columns
print("Unregistered users are the reason for NaN-values in userInputs columns:", check_user_inputs_nan_condition(user_inputs))

Unregistered users are the reason for NaN-values in userInputs columns: True


Now that we have identified the reason for the NaN values in the userInputs columns as being unregistered users, it implies that these NaN values carry meaningful information. Therefore, we cannot simply remove the rows, as doing so would discard valuable data about the unregistered users. We replace the NaN-values with the number -1. For the time-data we choose a timestamp way out of our given time frame (2018-2021) as the value of the placeholder: 01.01.2025.
For the numerical data, we check if we can use the number as a placeholder first.

In [406]:
# non-time data columns
columns_contain_nan = ["userID", "milesRequested", "paymentRequired", "minutesAvailable", "WhPerMile", "kWhRequested"]
time_data_contain_nan = ["modifiedAt", "requestedDeparture"]

# check if relevant columns already contain -1
contains_zero_value = (charging_set[columns_contain_nan] == -1).any()
print("user-data column contains a -1 value:\n", contains_zero_value)

user-data column contains a -1 value:
 userID              False
milesRequested      False
paymentRequired     False
minutesAvailable    False
WhPerMile           False
kWhRequested        False
dtype: bool


In [407]:
# fill nan-placeholder in relevant columns
charging_set[columns_contain_nan] = charging_set[columns_contain_nan].fillna(-1)

# fill placholder in time data
placeholder_datetime_gmt = 'Wed, 01 Jan 2025 00:00:00 GMT'
charging_set[time_data_contain_nan] = charging_set[time_data_contain_nan].fillna(placeholder_datetime_gmt)

As we saw earlier, there are stil 4087 missing values in the column "doneChargingTime"

In [408]:
doneChargingTime_missing_values = charging_set["doneChargingTime"].isna().sum()
doneChargingTime_values = charging_set["doneChargingTime"].count()
print(
    f"Relative share of missing values in the column 'doneChargingTime': {(doneChargingTime_missing_values / doneChargingTime_values) * 100:.2f}%")

Relative share of missing values in the column 'doneChargingTime': 6.71%


The definition of the feature "doneChargingTime" provided in the Team Assignment document is that doneChargingTime is the last Time when "the last non-zero current draw recorded". This means that "doneChargingTime" is the timestamp at which the EV actively draws energy from the charging hub for the last time.

We see that in 6.71% of all records, "doneChargingTime" does not have a value. We interpret this as the EV not being charged to 100% state of charge in the respective charging session.

However, we want to check if "doneChargingTime" is always within the time period of "connectionTime" and "disconnectTime".

In [409]:
# Convert time-related columns to datetime format
charging_set['doneChargingTime'] = pd.to_datetime(charging_set['doneChargingTime'])
charging_set['connectionTime'] = pd.to_datetime(charging_set['connectionTime'])
charging_set['disconnectTime'] = pd.to_datetime(charging_set['disconnectTime'])

# Check condition: connectionTime < doneChargingTime < disconnectionTime
condition = (
        charging_set['doneChargingTime'].notna() &
        (charging_set['doneChargingTime'] > charging_set['connectionTime']) &
        (charging_set['doneChargingTime'] < charging_set['disconnectTime'])
)


# Calculate the percentage of cases where the condition is not satisfied
total_cases = len(charging_set)
unsatisfied_cases = (~condition).sum()
percentage_unsatisfied = (unsatisfied_cases / total_cases) * 100

# Calculate time differences in hours
charging_set['time_diff_hours'] = charging_set.apply(
    lambda row: (row['doneChargingTime'] - row['disconnectTime']).total_seconds() / 60 if pd.notna(row['doneChargingTime']) else 0,
    axis=1
)
mean_time_diff_unsatisfied = charging_set.loc[~condition, 'time_diff_hours'].mean() if unsatisfied_cases > 0 else None

# Mean time difference for rows where the condition is satisfied
mean_time_diff_satisfied = charging_set.loc[condition, 'time_diff_hours'].mean()

# Display results
print("Results:")
if condition.all():
    print("doneChargingTime is always earlier than disconnectTime and later than connectionTime.")
else:
    print("There are cases where doneChargingTime does not satisfy the condition.")
    print(f"Percentage of cases where the condition is not satisfied: {percentage_unsatisfied:.2f}%")

print(f"\nMean time difference (condition satisfied): {mean_time_diff_satisfied:.2f} min")
if mean_time_diff_unsatisfied is not None:
    print(f"Mean time difference (condition not satisfied): {mean_time_diff_unsatisfied:.2f} min")
else:
    print("No rows where the condition is not satisfied.")

# Create a dataframe with cases where the condition is not satisfied
charging_set_condition_unsatisfied = charging_set[~condition]

# Display the dataframe with unsatisfied cases
print("\nRows where the condition is not satisfied:")
charging_set_condition_unsatisfied


Results:
There are cases where doneChargingTime does not satisfy the condition.
Percentage of cases where the condition is not satisfied: 13.08%

Mean time difference (condition satisfied): -184.39 min
Mean time difference (condition not satisfied): -1.31 min

Rows where the condition is not satisfied:


id            connectionTime  \
12     5e23b149f9af8b5fe4b973db 2020-01-02 15:04:38+00:00   
20     5e23b149f9af8b5fe4b973e3 2020-01-02 15:28:47+00:00   
25     5e23b149f9af8b5fe4b973e8 2020-01-02 15:42:05+00:00   
26     5e23b149f9af8b5fe4b973e9 2020-01-02 15:57:24+00:00   
33     5e23b149f9af8b5fe4b973f0 2020-01-02 16:34:35+00:00   
...                         ...                       ...   
65028  5d2fbdd3f9af8b4d0dd0d546 2019-07-01 19:20:31+00:00   
65030  5d2fbdd3f9af8b4d0dd0d548 2019-07-01 20:01:07+00:00   
65033  5d2fbdd3f9af8b4d0dd0d54b 2019-07-01 21:58:45+00:00   
65034  5d2fbdd3f9af8b4d0dd0d54c 2019-07-01 22:02:21+00:00   
65035  5d2fbdd3f9af8b4d0dd0d54d 2019-07-01 22:23:44+00:00   

                 disconnectTime          doneChargingTime  kWhDelivered  \
12    2020-01-02 22:08:39+00:00 2020-01-02 22:09:36+00:00        25.567   
20    2020-01-02 19:01:54+00:00 2020-01-02 19:02:51+00:00         7.417   
25    2020-01-02 21:58:45+00:00 2020-01-02 21:59:42+00:00        36.701   
26    2020-01-02 16:35:37+00:00 2020-01-02 16:36:34+00:00         3.689   
33    2020-01-02 18:49:41+00:00 2020-01-02 18:50:38+00:00         7.120   
...                         ...                       ...           ...   
65028 2019-07-02 00:16:32+00:00 2019-07-02 00:16:42+00:00        26.324   
65030 2019-07-02 00:32:26+00:00 2019-07-02 00:32:59+00:00        21.588   
65033 2019-07-02 00:39:48+00:00 2019-07-02 00:40:21+00:00        16.864   
65034 2019-07-02 00:58:50+00:00 2019-07-02 00:59:23+00:00        18.335   
65035 2019-07-02 02:03:36+00:00 2019-07-02 02:04:06+00:00        22.815   

                                    sessionID  siteID  spaceID    stationID  \
12     1_1_178_824_2020-01-02 15:04:38.051735       1  AG-1F07  1-1-178-824   
20     1_1_193_827_2020-01-02 15:28:46.685366       1  AG-1F02  1-1-193-827   
25     1_1_179_797_2020-01-02 15:42:05.217965       1  AG-3F23  1-1-179-797   
26     1_1_179_781_2020-01-02 15:57:23.951170       1  AG-3F31  1-1-179-781   
33     1_1_179_790_2020-01-02 16:34:34.999200       1  AG-3F19  1-1-179-790   
...                                       ...     ...      ...          ...   
65028  1_1_179_783_2019-07-01 19:20:30.955300       1  AG-3F29  1-1-179-783   
65030  1_1_179_800_2019-07-01 20:01:06.782562       1  AG-3F32  1-1-179-800   
65033  1_1_179_794_2019-07-01 21:58:44.571011       1  AG-3F20  1-1-179-794   
65034  1_1_191_807_2019-07-01 22:02:20.810735       1  AG-4F47  1-1-191-807   
65035  1_1_179_781_2019-07-01 22:23:32.496137       1  AG-3F31  1-1-179-781   

       userID  WhPerMile  kWhRequested  milesRequested  minutesAvailable  \
12      528.0      250.0         50.00           200.0             555.0   
20     1283.0      350.0         42.00           120.0             238.0   
25      474.0      400.0         32.00            80.0             532.0   
26      724.0      400.0          8.00            20.0              89.0   
33     2276.0      600.0         18.00            30.0             165.0   
...       ...        ...           ...             ...               ...   
65028   458.0      400.0          8.00            20.0             204.0   
65030  1479.0      275.0         19.25            70.0             286.0   
65033   364.0      400.0         40.00           100.0             234.0   
65034  2050.0      333.0         29.97            90.0             309.0   
65035  1626.0      200.0         34.00           170.0              99.0   

                          modifiedAt paymentRequired  \
12     Thu, 02 Jan 2020 15:04:49 GMT            True   
20     Thu, 02 Jan 2020 15:29:16 GMT            True   
25     Thu, 02 Jan 2020 15:42:11 GMT            True   
26     Thu, 02 Jan 2020 15:57:39 GMT            True   
33     Thu, 02 Jan 2020 16:35:06 GMT            True   
...                              ...             ...   
65028  Mon, 01 Jul 2019 19:23:09 GMT            True   
65030  Mon, 01 Jul 2019 20:03:21 GMT            True   
65033  Mon, 01 Jul 2019 21:59

So now we see that the doneChargingTime is not always in the time frame between connectionTime and disconnectTime. This occurs in 7.25% of the time. However, since the mean difference between disconnectTime and doneChargingTime is roughly 2.5 minutes, we assume that this imprecision is due to technical issues during the data collection (e.g. delay in recording the last draw by the charging hub). Therefore we set the values of "doneChargingTime" to be equal with "disconnectTime" for these records.  

In [410]:
# update doneChargingTime to disconnectTime for rows where the condition is not satisfied
charging_set.loc[~condition & charging_set['doneChargingTime'].notna(), 'doneChargingTime'] = charging_set.loc[~condition & charging_set['doneChargingTime'].notna(), 'disconnectTime']
charging_set[~condition][['doneChargingTime', 'disconnectTime']]


doneChargingTime            disconnectTime
12    2020-01-02 22:08:39+00:00 2020-01-02 22:08:39+00:00
20    2020-01-02 19:01:54+00:00 2020-01-02 19:01:54+00:00
25    2020-01-02 21:58:45+00:00 2020-01-02 21:58:45+00:00
26    2020-01-02 16:35:37+00:00 2020-01-02 16:35:37+00:00
33    2020-01-02 18:49:41+00:00 2020-01-02 18:49:41+00:00
...                         ...                       ...
65028 2019-07-02 00:16:32+00:00 2019-07-02 00:16:32+00:00
65030 2019-07-02 00:32:26+00:00 2019-07-02 00:32:26+00:00
65033 2019-07-02 00:39:48+00:00 2019-07-02 00:39:48+00:00
65034 2019-07-02 00:58:50+00:00 2019-07-02 00:58:50+00:00
65035 2019-07-02 02:03:36+00:00 2019-07-02 02:03:36+00:00

[8508 rows x 2 columns]

For the charging sessions where the car does not fully charge, we replace the NaN values with a placeholder just like we did for the userInputs. Here, we choose a time way out of our given time frame (2018-2021) as the value of the placeholder: 01.01.2025.

In [411]:
placeholder_datetime = pd.to_datetime('2025-01-01 00:00:00+00:00')
charging_set["doneChargingTime"] = charging_set["doneChargingTime"].fillna(placeholder_datetime)
charging_set[charging_set['doneChargingTime'] == placeholder_datetime]

id            connectionTime  \
3756   5ea3848df9af8b428bf97225 2020-04-08 16:36:53+00:00   
3757   5ea4d60df9af8b46573f1ee7 2020-04-09 13:23:52+00:00   
3758   5ea4d60df9af8b46573f1ee8 2020-04-09 13:40:03+00:00   
3759   5ea4d60df9af8b46573f1ee9 2020-04-09 14:01:53+00:00   
3760   5ea4d60df9af8b46573f1eea 2020-04-09 15:01:06+00:00   
...                         ...                       ...   
60396  5cb3d2a8f9af8b4551261e66 2019-03-29 16:54:49+00:00   
60411  5cb3d2a8f9af8b4551261e75 2019-03-30 00:32:10+00:00   
61097  5cc8e891f9af8b45d975b819 2019-04-14 16:19:12+00:00   
61103  5cca3a54f9af8b49aaa4cbac 2019-04-15 13:00:59+00:00   
61117  5cca3a54f9af8b49aaa4cbba 2019-04-15 13:52:45+00:00   

                 disconnectTime          doneChargingTime  kWhDelivered  \
3756  2020-04-09 03:43:29+00:00 2025-01-01 00:00:00+00:00        25.183   
3757  2020-04-09 17:59:02+00:00 2025-01-01 00:00:00+00:00        14.136   
3758  2020-04-09 21:43:22+00:00 2025-01-01 00:00:00+00:00        14.203   
3759  2020-04-09 16:33:40+00:00 2025-01-01 00:00:00+00:00        13.041   
3760  2020-04-09 23:38:08+00:00 2025-01-01 00:00:00+00:00         5.797   
...                         ...                       ...           ...   
60396 2019-03-29 21:55:22+00:00 2025-01-01 00:00:00+00:00        15.753   
60411 2019-03-30 02:23:14+00:00 2025-01-01 00:00:00+00:00        11.148   
61097 2019-04-14 20:38:39+00:00 2025-01-01 00:00:00+00:00        11.678   
61103 2019-04-15 13:55:25+00:00 2025-01-01 00:00:00+00:00         3.343   
61117 2019-04-15 14:47:01+00:00 2025-01-01 00:00:00+00:00         4.558   

                                    sessionID  siteID  spaceID    stationID  \
3756   1_1_194_822_2020-04-08 16:36:53.316264       1  AG-1F12  1-1-194-822   
3757   1_1_193_827_2020-04-09 13:23:51.689520       1  AG-1F02  1-1-193-827   
3758   1_1_178_823_2020-04-09 13:39:55.115497       1  AG-1F08  1-1-178-823   
3759   1_1_178_828_2020-04-09 14:01:53.467693       1  AG-1F10  1-1-178-828   
3760   1_1_193_820_2020-04-09 15:01:06.343454       1  AG-1F04  1-1-193-820   
...                                       ...     ...      ...          ...   
60396  1_1_179_799_2019-03-29 16:54:49.239801       1  AG-3F25  1-1-179-799   
60411  1_1_193_829_2019-03-30 00:32:10.128651       1  AG-1F03  1-1-193-829   
61097  1_1_193_816_2019-04-14 16:19:11.864262       1  AG-1F05  1-1-193-816   
61103  1_1_193_819_2019-04-15 13:00:58.722284       1  AG-1F06  1-1-193-819   
61117  1_1_179_800_2019-04-15 13:52:44.693153       1  AG-3F32  1-1-179-800   

       userID  WhPerMile  kWhRequested  milesRequested  minutesAvailable  \
3756    507.0      250.0          75.0           300.0             720.0   
3757    419.0      400.0          40.0           100.0             479.0   
3758   5065.0      400.0          24.0            60.0             483.0   
3759    651.0      400.0          20.0            50.0             183.0   
3760    466.0      300.0           6.0            20.0             182.0   
...       ...        ...           ...             ...               ...   
60396  1078.0      400.0          20.0            50.0             139.0   
60411   194.0      400.0          24.0            60.0             137.0   
61097   364.0      400.0          56.0           140.0             222.0   
61103   651.0      400.0          20.0            50.0             158.0   
61117   194.0      400.0          12.0            30.0              77.0   

                          modifiedAt paymentRequired  \
3756   Wed, 08 Apr 2020 16:37:01 GMT            True   
3757   Thu, 09 Apr 2020 13:24:06 GMT            True   
3758   Thu, 09 Apr 2020 13:40:03 GMT            True   
3759   Thu, 09 Apr 2020 14:02:09 GMT            True   
3760   Thu, 09 Apr 2020 15:01:20 GMT            True   
...                              ...             ...   
60396  Fri, 29 Mar 2019 16:55:03 GMT            True   
60411  Sat, 30 Mar 2019 00:32:57 GMT            True   
61097  Sun, 14 Apr 2019 16:19

## 1.5 Outliers 

In [422]:
# Check ranges of numeric columns
numeric_columns = charging_set.select_dtypes(include=[np.number])
for col in numeric_columns.columns:
    valid_values = numeric_columns.loc[numeric_columns[col] != -1, col]  # exclude -1, since we used it as a placeholder for NaN-values
    print(f"Column '{col}': min={valid_values.min()}, max={valid_values.max()}")

# Check unique values of categorical columns
categorical_columns = charging_set.select_dtypes(exclude=[np.number])
for col in categorical_columns.columns:
    print(f"Column '{col}': unique values={categorical_columns[col].unique()}")

Column 'kWhDelivered': min=0.501, max=108.79724166666666
Column 'siteID': min=1, max=2
Column 'userID': min=1.0, max=19923.0
Column 'WhPerMile': min=50.0, max=2000.0
Column 'kWhRequested': min=0.0, max=215.32
Column 'milesRequested': min=0.0, max=775.0
Column 'minutesAvailable': min=1.0, max=10062.0
Column 'time_diff_hours': min=-9367.283333333333, max=59.93333333333333
Column 'connectionTime_year': min=2018, max=2021
Column 'connectionTime_month': min=1, max=12
Column 'connectionTime_day': min=1, max=31
Column 'connectionTime_weekday': min=0, max=6
Column 'disconnectTime_year': min=2018, max=2021
Column 'disconnectTime_month': min=1, max=12
Column 'disconnectTime_day': min=1, max=31
Column 'disconnectTime_weekday': min=0, max=6
Column 'doneChargingTime_year': min=2018, max=2025
Column 'doneChargingTime_month': min=1, max=12
Column 'doneChargingTime_day': min=1, max=31
Column 'doneChargingTime_weekday': min=0, max=6
Column 'modifiedAt_year': min=2018, max=2025
Column 'modifiedAt_month'

We identify multiple columns with outliers:
    - (kWhDelivered: min approx. 2 km range)
- WhPerMile: energy consumption of 2000 too high
- minutesAvailable: 10.000 minutes = almost 7 days
- connectionDuration: min. 0.05h = 3 min
- chargingDuration: min. -0.68h, max = 200h

Let's have a look at the features that seem to have extreme outliers:

## 1.6 Feature Enigneering

### 1.5.1 Time Features

The time data is recorded in UTC format, which is not easily readable for humans. Therefore, we break it down into more human-readable components, such as year, month, day and time.

In [412]:
# convert remaining time data features
# connectionTime, disconnectTime and doneChargingTime have already been converted while handling missing data (chapter 1.4)
charging_set['modifiedAt'] = pd.to_datetime(charging_set['modifiedAt'])
charging_set['requestedDeparture'] = pd.to_datetime(charging_set['requestedDeparture'])

charging_set['connectionTime_year'] = charging_set['connectionTime'].dt.year
charging_set['connectionTime_month'] = charging_set['connectionTime'].dt.month
charging_set['connectionTime_day'] = charging_set['connectionTime'].dt.day
charging_set['connectionTime_time'] = charging_set['connectionTime'].dt.time
charging_set['connectionTime_weekday'] = charging_set['connectionTime'].dt.weekday

charging_set['disconnectTime_year'] = charging_set['disconnectTime'].dt.year
charging_set['disconnectTime_month'] = charging_set['disconnectTime'].dt.month
charging_set['disconnectTime_day'] = charging_set['disconnectTime'].dt.day
charging_set['disconnectTime_time'] = charging_set['disconnectTime'].dt.time
charging_set['disconnectTime_weekday'] = charging_set['disconnectTime'].dt.weekday

charging_set['doneChargingTime_year'] = charging_set['doneChargingTime'].dt.year
charging_set['doneChargingTime_month'] = charging_set['doneChargingTime'].dt.month
charging_set['doneChargingTime_day'] = charging_set['doneChargingTime'].dt.day
charging_set['doneChargingTime_time'] = charging_set['doneChargingTime'].dt.time
charging_set['doneChargingTime_weekday'] = charging_set['doneChargingTime'].dt.weekday

charging_set['modifiedAt_year'] = charging_set['modifiedAt'].dt.year
charging_set['modifiedAt_month'] = charging_set['modifiedAt'].dt.month
charging_set['modifiedAt_day'] = charging_set['modifiedAt'].dt.day
charging_set['modifiedAt_time'] = charging_set['modifiedAt'].dt.time
charging_set['modifiedAt_weekday'] = charging_set['modifiedAt'].dt.weekday

charging_set['requestedDeparture_year'] = charging_set['requestedDeparture'].dt.year
charging_set['requestedDeparture_month'] = charging_set['requestedDeparture'].dt.month
charging_set['requestedDeparture_day'] = charging_set['requestedDeparture'].dt.day
charging_set['requestedDeparture_time'] = charging_set['requestedDeparture'].dt.time
charging_set['requestedDeparture_weekday'] = charging_set['requestedDeparture'].dt.weekday

To better analyse the objectives later on we isolate valuable information into own features such as:
- connectionDuration: the time a vehicle is connected to the charging hub (in hours)
- chargingDuration: the time a vehicle is actively drawing energy from the charging hub (in hours)  

In [420]:
charging_set["connectionDuration"] = (charging_set["disconnectTime"] - charging_set["connectionTime"]).dt.total_seconds()/3600

charging_set["chargingDuration"] = charging_set.apply(
    lambda row: row["connectionDuration"] if row["doneChargingTime"] == placeholder_datetime
    else (row["doneChargingTime"] - row["connectionTime"]).total_seconds()/3600,
    axis=1
)

Since the exact ID assigned to a registered user is not of primary interest, but rather the higher-level information about whether a user is registered or not, we introduce the binary feature 'isRegistered' to indicate the user's registration status in the dataset.

This also applies to the feature "doneChargingTime". The main focus is not on the exact time of day when the car reaches 100% charge (i.e., doneChargingTime), but rather on whether the car has charged fully and, if it has, how long the car remained connected to the charging hub without further charging (i.e., how long it stayed plugged in after reaching 100%).

In [379]:
# isRegistered should be 0 if a user is not registered (userID == NaN) and be 1 if a user is registered
charging_set['isRegistered'] = (charging_set['userID'] != -1).astype(int)

# fullyCharged
charging_set['fullyCharged'] = (charging_set['doneChargingTime'] != placeholder_datetime).astype(int)

# parkingTime
charging_set['parkingDuration'] = charging_set.apply(
    lambda row: (row['disconnectTime'] - row['doneChargingTime']).total_seconds() / 3600
    if row['fullyCharged'] == 1 else 0, axis=1
)

# Optional: Round parkingTime to 2 decimal places for readability
charging_set['parkingDuration'] = charging_set['parkingDuration'].round(2)

# Display the result
charging_set[['doneChargingTime', 'disconnectTime', 'fullyCharged', 'parkingDuration']]

doneChargingTime            disconnectTime  fullyCharged  \
0     2020-01-02 17:31:35+00:00 2020-01-02 19:11:15+00:00             1   
1     2020-01-02 20:18:05+00:00 2020-01-02 22:38:21+00:00             1   
2     2020-01-02 16:35:06+00:00 2020-01-03 00:39:22+00:00             1   
3     2020-01-02 15:18:45+00:00 2020-01-02 16:38:39+00:00             1   
4     2020-01-02 18:17:30+00:00 2020-01-02 22:08:40+00:00             1   
...                         ...                       ...           ...   
65032 2019-07-01 23:14:33+00:00 2019-07-02 01:37:18+00:00             1   
65033 2019-07-02 00:39:48+00:00 2019-07-02 00:39:48+00:00             1   
65034 2019-07-02 00:58:50+00:00 2019-07-02 00:58:50+00:00             1   
65035 2019-07-02 02:03:36+00:00 2019-07-02 02:03:36+00:00             1   
65036 2019-07-02 00:24:32+00:00 2019-07-02 01:10:52+00:00             1   

       parkingDuration  
0                 1.66  
1                 2.34  
2                 8.07  
3                 1.33  
4                 3.85  
...                ...  
65032             2.38  
65033             0.00  
65034             0.00  
65035             0.00  
65036             0.77  

[65037 rows x 4 columns]